### PR TITLE
Fixed #31534 -- Deprecated django.conf.urls.url().

### PR DESCRIPTION
**Pull request by smithdc1**
[Ticket](https://code.djangoproject.com/ticket/31534)

### CODE REVIEW

File: django/conf/urls/__init__.py [hunk 1]
```diff
@@ -10,4 +13,9 @@
 
 
 def url(regex, view, kwargs=None, name=None):
+    warn(
+        'django.conf.urls.url is deprecated in favour of django.urls.re_path',
```
Comment (felixxm) [inline django/conf/urls/__init__.py:17]: `django.conf.urls.url` -> `django.conf.urls.url()`
`django.urls.re_path` -> `django.urls.re_path()`

File: django/conf/urls/__init__.py [hunk 2]
```diff
@@ -10,4 +13,9 @@
...
+        RemovedInDjango40Warning,
+        stacklevel=2
```
Comment (felixxm) [inline django/conf/urls/__init__.py:19]: Why `stacklevel=2`? :thinking: IMO we can remove it.

File: docs/internals/deprecation.txt [hunk 3]
```diff
@@ -81,6 +81,9 @@ details on these changes.
 * The model ``NullBooleanField`` will be removed. A stub field will remain for
   compatibility with historical migrations.
 
+* ``django.conf.urls.url`` an alias for ``django.urls.re_path`` will be
+  removed.
```
Comment (felixxm) [inline docs/internals/deprecation.txt:85]:
```
* ``django.conf.urls.url()`` will be removed.
```

File: docs/releases/3.1.txt [hunk 4]
```diff
@@ -475,6 +475,9 @@ URLs
 * :ref:`Path converters <registering-custom-path-converters>` can now raise
   ``ValueError`` in ``to_url()`` to indicate no match when reversing URLs.
 
+* :func:`django.conf.urls.url` is deprecated use :func:`django.urls.re_path`
```
Comment (felixxm) [inline docs/releases/3.1.txt:478]: Maybe:
```
* ``django.conf.urls.url()`` alias of :func:`django.urls.re_path` is deprecated.
```

File: docs/ref/urls.txt [hunk 5]
```diff
@@ -144,6 +144,10 @@ Helper function to return a URL pattern for serving files in debug mode::
 This function is an alias to :func:`django.urls.re_path()`. It's likely to be
 deprecated in a future release.
 
+.. deprecated:: 3.1
+
+    ``django.conf.urls.url`` is deprecated use ``django.urls.re_path`` instead.
```
Comment (felixxm) [inline docs/ref/urls.txt:149]: Maybe:
```
.. deprecated:: 3.1

    Alias of :func:`django.urls.re_path` for backwards compatibility.
```

File: tests/urlpatterns/tests.py [hunk 6]
```diff
@@ -303,3 +307,21 @@ def raises_type_error(value):
...
+
+
+class DeprecationTests(TestCase):
+    @ignore_warnings(category=RemovedInDjango40Warning)
+    def test_conf_url(self):
+        urls = (
+            conf_url(r'^regex/(?P<pk>[0-9]+)/$', empty_view, name='regex'),
+            re_path(r'^regex/(?P<pk>[0-9]+)/$', empty_view, name='regex'),
+        )
+        assert str(urls[0]) == str(urls[1])
```
Comment (felixxm) [inline tests/urlpatterns/tests.py:319]: Use `assertEqual()`, e.g.
```python
self.assertEqual(
    conf_url(r'^regex/(?P<pk>[0-9]+)/$', empty_view, name='regex'),
    re_path(r'^regex/(?P<pk>[0-9]+)/$', empty_view, name='regex'), 
)
```

File: django/conf/urls/__init__.py (hunk 1 above)
Comment (pope1ni) [reply django/conf/urls/__init__.py:17]: Also `favour` → `favor`.

File: tests/urlpatterns/tests.py (hunk 6 above)
Comment (smithdc1) [reply tests/urlpatterns/tests.py]: @felixxm `URLPattern` doesn't have an equality function. I've used `assertEqual()` but converted them to strings first. Is this ok?

`AssertionError: <URLPattern '^regex/(?P<pk>[0-9]+)/$' [name='regex']> != <URLPattern '^regex/(?P<pk>[0-9]+)/$' [name='regex']>`

Comment (felixxm) [reply tests/urlpatterns/tests.py]: TBH, I don't think we need this test. I will chop it.

**Comment by felixxm** [comment]
@smithdc1 Thanks :+1: I pushed small edits.

File: django/conf/urls/__init__.py (hunk 2 above)
Comment (jdufresne) [reply django/conf/urls/__init__.py]: I added `stacklevel=2` back in #12960. IMO, it assists the library user with identifying the line they should modify. Without the argument, the warning points to Django code, but with the argument, the warning points to the user's code that much change. There is more detail in the PR too.

Comment (felixxm) [reply django/conf/urls/__init__.py]: True, thanks :+1:

**Pull request merged**